core/src/main/scala/org/apache/spark/sql/execution/columnar/impl/ColumnFormatRelation.scala

```diff
@@ -417,8 +422,90 @@ class ColumnFormatRelation(
     })
     ColumnFormatRelation.flushLocalBuckets(resolvedName)
   }
+
+  /**
+    * Index table is same as the column table apart from how it is
+    * partitioned and colocated. Add GEM_PARTITION_BY and GEM_COLOCATE_WITH
+    * clause in its options. Also add GEM_INDEXED_TABLE parameter to
+    * indicate that this is an index table.
+    */
+  private def createIndexTable(indexIdent: QualifiedTableName,
+                               tableIdent: QualifiedTableName,
+                               tableRelation: JDBCAppendableRelation,
+                               indexColumns: Seq[String],
+                               options: Map[String, String]): Unit = {
+
+
+    val parameters = new CaseInsensitiveMutableHashMap(options)
+    val snc = _context.asInstanceOf[SnappyContext]
+    val indexTblName = snc.getIndexTable(indexIdent).toString
+    val caseInsensitiveMap = new CaseInsensitiveMutableHashMap(tableRelation.origOptions)
+    val tempOptions = caseInsensitiveMap.filterNot(pair => {
+      pair._1.equals(Utils.toLowerCase(StoreUtils.PARTITION_BY)) ||
+        pair._1.equals(Utils.toLowerCase(StoreUtils.COLOCATE_WITH)) ||
+        pair._1.equals(Utils.toLowerCase(JdbcExtendedUtils.DBTABLE_PROPERTY)) ||
+        pair._1.equals(Utils.toLowerCase(ExternalStoreUtils.INDEX_NAME))
+    }).toMap + (StoreUtils.PARTITION_BY -> indexColumns.mkString(",")) +
+      (StoreUtils.GEM_INDEXED_TABLE -> tableIdent.toString) +
+      (JdbcExtendedUtils.DBTABLE_PROPERTY -> indexTblName)
+
+    val indexOptions = parameters.get(StoreUtils.COLOCATE_WITH) match {
+      case Some(value) => tempOptions + (StoreUtils.COLOCATE_WITH -> value)
+      case _ => tempOptions
+    }
+    snc.createTable(
+      indexTblName,
+      "column",
+      tableRelation.schema,
+      indexOptions)
+  }
+
```

Comment:
I am assuming you plan on adding local CachedBatch sorting and merge iterator later. In fact it can be a general table level thing (so table can also be sorted per CachedBatch on the partitioning column) using a common property. For the merge iterator we may need some hint from the higher level that this is from sorting or not else it will be unnecessary merge sorting.